geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/IngestCommand.scala

```diff
@@ -231,107 +238,63 @@ object IngestCommand extends LazyLogging {
-        def openStdin(): InputStream = {
-          try {
-            // Get the CachingStdInHandle that gives us access to our custom-made input stream
-            val handle = inputs.handles.headOption.getOrElse {
-              throw new RuntimeException("Unable to obtain handle for stdin")
-            }
-
-            // Use the handle to get the stream
-            val (_, stream) = handle.open.next()
-            opened += CloseableIterator.single(stream)
-            stream
-          } catch {
-            case e: Exception => throw new RuntimeException("Unable to read from stdin", e)
-          }
-        }
-
-        open = openStdin
+    if (converter == null && (inputs.stdin || inputs.paths.nonEmpty)) {
+      val errorMsg = if (inputs.stdin) { "Unable to read data from stdin" } else {
+        "Parameter <files> did not evaluate to something that could be read"
       }
-      // try to infer the schema from the input files themselves
-      else {
-        Command.user.info("No converter defined - will attempt to detect schema from input files")
-        file = inputs.handles.headOption.getOrElse {
-          throw new ParameterException("Parameter <files> did not evaluate to anything that could be read")
-        }
-
-        def openFile(): InputStream = {
+      // try to infer the schema from the inputs
+      Command.user.info(
+        s"No converter defined - will attempt to detect schema from ${if (inputs.stdin) "stdin" else "input files"}")
+      val file = inputs.handles.headOption.getOrElse(throw new ParameterException(errorMsg))
+      val opened = ListBuffer.empty[CloseableIterator[InputStream]]
+      def open(): InputStream = {
+        val is = try {
           val streams = file.open.map(_._2)
           opened += streams
-          if (streams.hasNext) {
-            streams.next
-          } else {
-            throw new ParameterException("Parameter <files> did not evaluate to anything that could be read")
-          }
+          if (streams.hasNext) { streams.next } else { null }
+        } catch {
+          case NonFatal(e) => throw new RuntimeException(errorMsg, e)
         }
-
-        open = openFile
-      }
-
-      val (inferredSft, inferredConverter) = try {
-        val opt = format match {
-          case None => if (inputs.paths == Inputs.StdInInputs) {
-              SimpleFeatureConverter.infer(open, Option(sft))
-            } else {
-              SimpleFeatureConverter.infer(open, Option(sft), Option(file.path))
-            }
-
-          case Some(fmt) => if (inputs.paths == Inputs.StdInInputs) {
-            TypeAwareInference.infer(fmt, open, Option(sft), None)
-          } else {
-            TypeAwareInference.infer(fmt, open, Option(sft), Option(file.path))
-          }
+        if (is == null) {
+          throw new ParameterException(errorMsg)
+        } else {
+          is
```

Comment:
I'm thinking that we should change this to something like:
```scala
is.getOrElse({
    throw new ParameterException(errorMsg)
})
```

and change the `open()` function's return type to `Option[InputStream]` to avoid having to return `null` and possibly risking a `NullPointerException`.

Reply:
open has to return an Iterator, not an option, in order for us to pass it to the type inference method. i split up the logic this way so that we could distinguish between exceptions opening the stream (runtime exceptions) vs exceptions where there was no input (parameter exception). another option would have been to throw an exception `if (!streams.hasNext)`, then catch and rethrow the parameter exception, but that seemed more complicated to me. we could make `is` be an Option inside the function, and then use the `getOrElse` as you outlined, but since it's a small localized function the null is easy to reason about, and doesn't leak out, so I'm ok with it. if you feel strongly about it, feel free to change it, though.

Reply:
Makes sense. I don't feel too strongly about it, so we can keep it as is!